src/transformers/trainer.py

```diff
@@ -373,6 +397,143 @@ def __init__(
+        optimizer = "Adam"  # currently trainer only has Adam config params
+
+        # override only if the ds config doesn't already have this section
+        if "optimizer" in config:
+            logger.info("Keeping the optimizer config from the config file intact")
+        else:
+            config["optimizer"] = {
+                "type": optimizer,
+                "params": optimizer_configs[optimizer],
+            }
+
+        # DS schedulers (deepspeed/runtime/lr_schedules.py):
+        #
+        # DS name      | --lr_scheduler_type  | HF func                           | Notes
+        # -------------| ---------------------|-----------------------------------|--------------------
+        # LRRangeTest  | na                   | na                                | LRRT
+        # OneCycle     | na                   | na                                | 1CLR
+        # WarmupLR     | constant_with_warmup | get_constant_schedule_with_warmup | w/ warmup_min_lr=0
+        # WarmupDecayLR| linear               | get_linear_schedule_with_warmup   |
+
+        # override only if the ds config doesn't already have this section
+        if "scheduler" in config:
+            logger.info(
+                "Keeping the scheduler config from the config file intact, ignoring any scheduler-specific cl args"
+            )
+        else:
+            if args.lr_scheduler_type == SchedulerType.LINEAR:
+                scheduler = "WarmupDecayLR"
+                params = {
+                    "last_batch_iteration": -1,
+                    "total_num_steps": self.get_num_training_steps(),
+                    "warmup_min_lr": 0,
+                    "warmup_max_lr": args.learning_rate,
+                    "warmup_num_steps": args.warmup_steps,
+                }
+            elif args.lr_scheduler_type == SchedulerType.CONSTANT_WITH_WARMUP:
+                scheduler = "WarmupLR"
+                params = {
+                    "warmup_min_lr": 0,
+                    "warmup_max_lr": args.learning_rate,
+                    "warmup_num_steps": args.warmup_steps,
+                }
+            else:
+                raise ValueError(f"{args.lr_scheduler_type} scheduler type is not supported by DeepSpeed")
+
+            config["scheduler"] = {
+                "type": scheduler,
+                "params": params,
+            }
+
+        # fp16
+        if "fp16" in config:
+            logger.info("Keeping the fp16 config from the config file intact, ignoring any fp16-specific cl args")
+        elif args.fp16:
+            # XXX: not sure what to do with args.fp16_backend
+            config["fp16"] = {
+                "enabled": True,
+                "opt_level": args.fp16_opt_level,
+            }
+
+        # init that takes some config via `args`, and the bulk of it via `config_params`
+        model_parameters = filter(lambda p: p.requires_grad, model.parameters())
+        model, optimizer, _, lr_scheduler = deepspeed.initialize(
```

Comment:
@stas00 Suppose someone wanted to use both the current HF model-parallel feature for GPT-2/T5 with deepspeed. How would that work?

Reply:
Once I'm done with the integration this is the next thing on my todo list - see how we can integrate our currently basic version of MP with DeepSpeed and also PipelineParallelism tools.

According to the Deepspeed paper it should just work, so you can go ahead and try it already and report back.

The key message the Deepspeed paper communicates as I understand it that unless you have some insanely big model you won't need MP.

Reply:
Yeah you won't really need MP after DeepSpeed Stage 3 comes out, except in certain scenarios like wanting to use activation partitioning.

Reply:
That's exciting! When is that scheduled for?

Reply:
No idea about ETA lol, last I checked, Shaden said they were working on it. I'm guessing definitely this spring!

Reply:
Awesome!

stage 3 is "Parameter Partitioning (Pos+g+p )", correct? I had no idea the paper wasn't already fully implemented. 

Thank you for that insight, @g-karthik

Reply:
Yes, Stage 3 is parameter partitioning.

Reply:
So I gave it a quick try and it works just fine with ds and mp (using t5)

```
export BS=20; rm -r output_dir; CUDA_VISIBLE_DEVICES=0,1 PYTHONPATH=../../src USE_TF=0 deepspeed  ./finetune_trainer.py --model_name_or_path t5-small --output_dir output_dir --adam_eps 1e-06 --data_dir wmt_en_ro --do_eval --do_predict --do_train --evaluation_strategy=steps --freeze_embeds --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 1000 --max_source_length 128 --max_target_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size $BS --per_device_train_batch_size $BS --predict_with_generate --eval_steps 25000  --sortish_sampler --task translation_en_to_ro --test_max_target_length 128 --val_max_target_length 128 --warmup_steps 500 --n_train 200 --n_val 200 --n_test 200 --deepspeed ds_config.json --fp16 --save_steps 2 --model_parallel
```

need to study/compare quality/gpu utilization with some bigger number of samples

Reply:
Interesting, my understanding is that DeepSpeed internally places the model to a CUDA device. So I was expecting that the current explicit device map placement approach to MP would cause things to go haywire. Will read more to understand what's going on under the hood.

Reply:
hmm, I don't think `finetune_trainer.py` actually does anything on my branch - perhaps I somehow dropped it - let me check.

Reply:
yeah, no, the original MP PR didn't actually enable MP in trainer: https://github.com/huggingface/transformers/pull/8696
at least I don't see any `parallelize()` calls. So everything is prepared but no MP.

@sgugger, am I missing something? Was `model.parallelize()` ever called inside `trainer.py`?

Bottom line, the trainer isn't running MP so my test wasn't testing what I thought it was testing.

Reply:
Unless I'm missing something it doesn't look it was ever put in or tested it to work, I added:

```
        if self.args.model_parallel:
            if model.is_parallelizable:
                model.parallelize()
            else:
                raise ValueError(
                    f"{model.__class__.__name__} implementation currently doesn't support model parallelism, therefore --model_parallel cl arg cannot be used"
                )

```
and tried to run it on 2 gpus:
```
rm -r output_dir; CUDA_VISIBLE_DEVICES=0,1 PYTHONPATH=../../src USE_TF=0 ./finetune_trainer.py --model_name_or_path t5-small --output_dir output_dir --adam_eps 1e-06 --data_dir wmt_en_ro --do_eval --do_train --evaluation_strategy=steps --fp16 --freeze_embeds --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 1000 --max_source_length 128 --max_target_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size 4 --per_device_train_batch_size 4 --predict_with_generate --eval_steps 25000 --save_steps 25000 --sortish_sampler --task translation_en_to_ro --test_max_target_length 128 --val_max_target_length 128 --warmup_steps 1 --n_train 2 --n_val 2 --n_test 2 --do_predict --model_parallel --model_parallel
```
it fails `RuntimeError: Input, output and indices must be on the current device` - it doesn't have the setup right.

So I take my words back - you can't test DS with MP at the moment, since the trainer/MP isn't working - at least on my branch.

Reply:
@alexorona, was there an MP activation in the trainer? Did I/we somehow loose it? for some reason I can't find it and if I activate it with the code above it doesn't work.

I have never used trainer/MP and was using my own code to test MP, so I had no idea there was an issue there.